File: concrete-agent/packages/core-backend/app/mcp/tools/element_name_normalizer.py
"""
Element-name normalization layer (W3).

A single pure function that runs **immediately before** element classification and
rewrites a raw element name into three signals the classifier consumes:

  - canonical_name        — head-noun canonicalized + prepositional/participle
                            tail stripped + dimensions removed, so the existing
                            (deliberately unchanged) keyword rule table matches on
                            the governing structural noun, not a stray substring.
  - construction_context  — 'bridge' | 'retaining_wall' | 'building' | None,
                            derived from object name + content keywords, DECOUPLED
                            from the assumption that a numbered object code (SO\\d{3})
                            means bridge.
  - status                — 'nový' | 'stávající', from existing/demolition wording.

No LLM. Deterministic (required for replay). The classifier and its rule table
stay a pure category matcher; this layer only annotates + rewrites the string the
matcher sees.

Reference: docs/tasks/TASK_W3_NormalizeElementName_SO250.md §2, §3.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional

# ── construction context ─────────────────────────────────────────────────────

# Genuine bridge vocabulary (name + object_code), NOT the bare SO-number. A
# numbered object code alone no longer implies bridge (task §3 / criterion #69).
_BRIDGE_CONTENT = re.compile(
    r"\bmost\w*|přemost|lávk|mostovk|nosn[áéí]\w*\s*konstr|\bNK\b|pilíř|"
    r"opěr[ay]\b|\bP\d|estakád|silnič.*most",
    re.I,
)

# Retaining / angular / abutment-style WALL vocabulary.
_WALL_CONTENT = re.compile(
    r"zárubní|opěrn[áé]\s*z[eě]ď|úhlov[áé]|tížn[áé]|gabion|opěrn[áé]\s*konstr",
    re.I,
)

# ── status (nový vs stávající) ───────────────────────────────────────────────

_EXISTING_STATUS = re.compile(
    r"stávajíc|demol|bourán|odstran|původní|stav\.\s|odbour", re.I
)

# ── modifier stripping (prepositional / participle tails) ────────────────────
# These introduce a subordinate ref (material, anchoring, purpose) — NOT the head
# noun. Cut the name at the first such marker so the head NP remains (task §2).
# Conservative, curated set (the task's own examples) to avoid over-stripping.
_TAIL_MARKERS = re.compile(
    r"\s+(?:z|ze)\s+\w"          # "z lomového kamene"
    r"|\s+kotven\w*"              # "kotvený do dříku"
    r"|\s+obložen\w*"            # "obložen lomovým kamenem"
    r"|\s+vlepen\w*"             # "vlepenými kotvami"
    r"|\s+pro\s+založen"         # "pro založení"
    r"|\s+na\s+líci",            # "na líci"
    re.I,
)

# Dimensions / numeric noise: "0,56×2,75", "0,56 x 2,75", standalone numbers.
_DIMENSIONS = re.compile(r"\d+[\.,]?\d*\s*[x×]\s*\d+[\.,]?\d*|\b\d+[\.,]\d+\b")


@dataclass(frozen=True)
class NormalizedName:
    """Output of the normalization layer, consumed by the classifier."""

    canonical_name: str          # string the rule table matches on
    construction_context: Optional[str]  # 'bridge' | 'retaining_wall' | 'building' | None
    status: str                  # 'nový' | 'stávající'
    original_name: str
    head_noun: Optional[str] = None  # canonical head, when one was resolved


def _strip_modifiers(name: str) -> str:
    """Cut the prepositional/participle tail + dimensions, leaving the head NP."""
    m = _TAIL_MARKERS.search(name)
    core = name[: m.start()] if m else name
    core = _DIMENSIONS.sub(" ", core)
    return re.sub(r"\s+", " ", core).strip()


def _construction_context(name: str, object_code: Optional[str]) -> Optional[str]:
    hay = f"{name} {object_code or ''}"
    # Wall wins over bridge when an explicit retaining-wall term is present
    # (a zárubní/úhlová zeď that mentions 'opěra' is still a wall, not a bridge).
    if _WALL_CONTENT.search(hay):
        return "retaining_wall"
    if _BRIDGE_CONTENT.search(hay):
        return "bridge"
    return None


def _canonical_head(core: str, context: Optional[str]) -> tuple[str, Optional[str]]:
    """Resolve the governing structural noun → a canonical token the EXISTING
    rule table already matches. Ordered: multi-word structural terms first, then
    context-sensitive 'dřík', then single nouns. Returns (canonical_name, head)."""
    low = core.lower()

    # 1. Nosná konstrukce / NK — governing concept beats a 'trám' modifier (#68).
    if re.search(r"nosn[áéí]\w*\s*konstr|\bNK\b", low):
        return "nosná konstrukce", "nosná konstrukce"

    # 2. Lícové zdivo / obklad — the new masonry-cladding category (#65).
    if re.search(r"lícov\w*\s*(?:obklad|zdiv)|obklad|kamenn\w*\s*zdiv", low):
        return "lícový obklad zdivo", "obklad"

    # 3. 'Dřík' is context-sensitive (#63 vs #64):
    #    bridge/pier context → pier shaft; otherwise → retaining-wall stem.
    if re.search(r"dřík", low) and not re.search(r"pilíř", low):
        if context == "bridge":
            return "dřík pilíře", "dřík"
        return "opěrná zeď", "opěrná zeď"

    # 4. Explicit pier.
    if re.search(r"pilíř", low):
        return "dřík pilíře", "pilíř"

    # 5. 'Základ' canonized to the plural the brittle rule expects (#66).
    #    (foundation slab handled by the existing 'základová deska' rule.)
    if re.search(r"základ", low) and not re.search(r"základov\w*\s*desk", low):
        return "základy", "základ"

    # No canonical rewrite — let the matcher run on the stripped core.
    return core, None


def normalize_element_name(
    name: str, object_code: Optional[str] = None
) -> NormalizedName:
    """Normalize a raw element name for classification. Pure + deterministic."""
    raw = name or ""
    status = "stávající" if _EXISTING_STATUS.search(raw) else "nový"
    context = _construction_context(raw, object_code)
    core = _strip_modifiers(raw)
    canonical, head = _canonical_head(core, context)
    return NormalizedName(
        canonical_name=canonical or core or raw,
        construction_context=context,
        status=status,
        original_name=raw,
        head_noun=head,
    )

File: Monolit-Planner/shared/src/classifiers/element-name-normalizer.ts
/**
 * Element-name normalizer — head-noun pre-classification layer (TASK_2b Gate 2b).
 *
 * Mirrors the W3 (Python) `element_name_normalizer.py`: a pure, deterministic
 * pass that runs BEFORE type selection and rewrites a raw element name into the
 * canonical head-noun string the keyword matcher should see — so a word inside a
 * prepositional / participle tail ("…kotvený do dříku", "Trámy nosné konstrukce")
 * does not decide the type. Head-noun is an ALGORITHM, never a lookup table.
 *
 * SELECTIVE port (per TASK_2b condition: the engine may be MORE precise than W3
 * and must not regress). Ported: NK-beats-modifier, context-sensitive dřík
 * (wall vs pier), základ canonicalization. NOT ported: W3's unconditional
 * "pilíř → pier" rule — the engine already distinguishes a building pilíř
 * (→ sloup) from a bridge pier (→ driky_piliru via bridge-remap), which is finer
 * than W3. The masonry-`obklad` reject head is materialized in Gate 3.
 *
 * Tail markers come from the single KB source (dictionaries.cs.tail_markers).
 */
import { ELEMENT_CLASSIFICATION_RULES } from '../kb-generated/element-classification-rules.js';

export type ConstructionContext = 'bridge' | 'retaining_wall' | 'building';

const TAIL_MARKERS: readonly string[] = ELEMENT_CLASSIFICATION_RULES.dictionaries.cs.tail_markers;
// Dimensions / numeric noise: "0,56×2,75", "0,56 x 2,75", standalone decimals.
const DIMENSIONS = /\d+[.,]?\d*\s*[x×]\s*\d+[.,]?\d*|\b\d+[.,]\d+\b/g;

export interface NormalizedName {
  /** String the keyword matcher should run on. */
  canonical: string;
  /** True when a head-noun rule rewrote the string (else canonical == stripped core). */
  headFired: boolean;
  /** The resolved governing noun, when one fired. */
  head?: string;
}

/** Diacritic-stripped lowercase, matching the classifier's `normalize`. */
function fold(text: string): string {
  return text.toLowerCase().normalize('NFD').replace(/[̀-ͯ]/g, '');
}

/** Cut the prepositional/participle tail (first KB tail marker) + dimensions. */
function stripModifiers(name: string): string {
  const low = name.toLowerCase();
  let cut = -1;
  for (const marker of TAIL_MARKERS) {
    const idx = low.indexOf(marker.toLowerCase());
    if (idx >= 0 && (cut === -1 || idx < cut)) cut = idx;
  }
  const core = cut >= 0 ? name.slice(0, cut) : name;
  return core.replace(DIMENSIONS, ' ').replace(/\s+/g, ' ').trim();
}

/**
 * Resolve the governing structural noun → a canonical string the EXISTING
 * keyword table already matches. Ordered: NK first (governing concept beats a
 * 'trám' modifier), then context-sensitive 'dřík', then 'základ'.
 */
export function normalizeElementName(name: string, context?: ConstructionContext): NormalizedName {
  const core = stripModifiers(name || '');
  const low = fold(core);

  // 1. Nosná konstrukce / NK — the head concept, beats a 'trám' modifier (#68).
  if (/nosn\w*\s*konstr/.test(low) || /\bnk\b/.test(low)) {
    return { canonical: 'nosná konstrukce', headFired: true, head: 'nosná konstrukce' };
  }

  // 2. Lícové zdivo / obklad z lomového kamene — masonry CLADDING, the head noun
  //    (#65). Resolved on the tail-stripped core so "…kotvený do dříku" cannot
  //    masquerade as a pier. head='obklad' signals a REJECT to the classifier
  //    (not a concrete structural element) — materialized in Gate 3.
  if (/licov\w*\s*(?:obklad|zdiv)/.test(low) || /\bobklad\b/.test(low) || /kamenn\w*\s*zdiv/.test(low) || /lomov\w*\s*kam/.test(low)) {
    return { canonical: 'lícový obklad zdivo', headFired: true, head: 'obklad' };
  }

  // 2. 'Dřík' is context-sensitive (#63 / #73 / #74): a pier shaft in a bridge,
  //    a retaining-wall stem otherwise. Guarded by "no pilíř" so explicit
  //    "dříky pilířů" keeps flowing through the normal keyword path (and the
  //    engine's finer pilíř handling). A co-occurring opěr is left for Gate 4.
  if (/drik/.test(low) && !/pilir/.test(low)) {
    return context === 'bridge'
      ? { canonical: 'dřík pilíře', headFired: true, head: 'dřík' }
      : { canonical: 'opěrná zeď', headFired: true, head: 'dřík' };
  }

  // NOTE — deliberately NO "pilíř → pier" head rule here. W3's normalizer maps
  // any 'pilíř' to a pier shaft unconditionally; porting that would make the
  // ENGINE LESS precise, because the engine already distinguishes a building
  // pilíř (→ sloup) from a bridge pier (→ driky_piliru via the bridge_remap)
  // using context. "Beton pilířů C30/37" with no context must stay sloup; only
  // a bridge context promotes it. So 'pilíř' flows untouched through the normal
  // keyword path (sloup) + bridge-remap — the engine being finer than W3, which
  // the directed family roll-up explicitly permits (engine may be more precise,
  // never cross a foreign family). Do NOT add a pilíř branch.

  // 3. 'Základ' (but not 'základová deska') → the plural the keyword rule
  //    expects (#66), recovering the head past a brittle suffix. Guarded by
  //    "no oper/pilir": "základ opěry"/"základ pilíře" carry a genitive
  //    qualifier (not a strippable tail) that the finer zaklady_oper /
  //    zaklady_piliru keyword rules need — collapsing to bare "základy" would
  //    drop it. Those names flow through the normal keyword path unchanged.
  if (
    /zaklad/.test(low) &&
    !/zakladov\w*\s*desk/.test(low) &&
    !/oper/.test(low) &&
    !/pilir/.test(low)
  ) {
    return { canonical: 'základy', headFired: true, head: 'základ' };
  }

  return { canonical: core, headFired: false };
}
